In [1]:
# Import the libraries that we will need in this project
import pandas as pd
import datetime as dt
import numpy as np
import requests
import tweepy
import json
import re
import time

# Introduction

Wrangling is an important skill evry data scientist/analyst must possess. We need to wrangle our data for optimum output. Wrangling, though a time consuming process, saves a lot of time while visualizing, analysing and creating predictive models. 
For this project, I will be gathering data from many sources in a variety of formats. Beyond this, I will asssess its quality and tidiness. Finally, I will clean this based on the definitions.

# Gather
I gathered data from three sources for this project as described below:
- Manually from the WeRateDogs twitter archive by clicking the following link: twitter_archive_enhanced.csv
- Programatically, from the tweet image predictions using requests library from this url: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv
- From Twitter API

In [2]:
# Read the twitter-archive-enhanced.csv file and store it as dataframe in archive
archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# Check the contents and structure of the dataframe
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [10]:
# Using Requests library to download a file then store it in a tsv file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open (url.split('/')[-1], mode = 'wb') as file:
    file.write(response.content)

# Read the downloaded file into a dataframe 'images'
images = pd.read_csv('image-predictions.tsv', sep='\t')
# Check the contents and structure
images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [9]:
# Authenticating for using twitter API
consumer_key = '0v8m71SMwKoSnOECEN5ChMkn6'
consumer_secret = "8QYwYcalgbACIt9HbqrYNEaDVoIChd8e44XySYp2BXPPzvwoGt"
access_token = '547075203-GKXJRAn0dNgCErEooBsIzAnMRdS2ZOFd1JT6gvXg'
access_token_secret = '3uKwH5ztXjbfH1SYJ4SoNOvZPBvy4tsPAgerWOKCXfmi8'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)

api = tweepy.API(auth, parser=tweepy.parsers.JSONParser(), #Parse the result to json object
                 wait_on_rate_limit=True) # Wait for rate limits to replenish

In [6]:
# List to store the dictionaries of our result
df_list = []
# List to store tweet_id of errors
error_list = []

# Calculate the time of excution
start = time.time()

# Get the tweet object for all the tweets in the archive dataframe 
tweet_id = images['tweet_id']
for id in tweet_id:
    try:
        page = api.get_status(id)
#         print ('hello')
        favorite_count = page['favorite_count']
        retweet_count = page['retweet_count']
        date_time= pd.to_datetime(page['created_at'])
        df_list.append({'tweet_id': int(id),
                  'favorite_count': int(favorite_count),
                  'retweet_count': int(retweet_count),
                  'date_time': pd.to_datetime(date_time)})
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(id)
#         print ('hi')
        
# Calculate the time of excution
end = time.time()
print(end - start)

0       666020888022790149
1       666029285002620928
2       666033412701032449
3       666044226329800704
4       666049248165822465
5       666050758794694657
6       666051853826850816
7       666055525042405380
8       666057090499244032
9       666058600524156928
10      666063827256086533
11      666071193221509120
12      666073100786774016
13      666082916733198337
14      666094000022159362
15      666099513787052032
16      666102155909144576
17      666104133288665088
18      666268910803644416
19      666273097616637952
20      666287406224695296
21      666293911632134144
22      666337882303524864
23      666345417576210432
24      666353288456101888
25      666362758909284353
26      666373753744588802
27      666396247373291520
28      666407126856765440
29      666411507551481857
               ...        
2045    886366144734445568
2046    886680336477933568
2047    886736880519319552
2048    886983233522544640
2049    887101392804085760
2050    887343217045368832
2

0       666020888022790149
1       666029285002620928
2       666033412701032449
3       666044226329800704
4       666049248165822465
5       666050758794694657
6       666051853826850816
7       666055525042405380
8       666057090499244032
9       666058600524156928
10      666063827256086533
11      666071193221509120
12      666073100786774016
13      666082916733198337
14      666094000022159362
15      666099513787052032
16      666102155909144576
17      666104133288665088
18      666268910803644416
19      666273097616637952
20      666287406224695296
21      666293911632134144
22      666337882303524864
23      666345417576210432
24      666353288456101888
25      666362758909284353
26      666373753744588802
27      666396247373291520
28      666407126856765440
29      666411507551481857
               ...        
2045    886366144734445568
2046    886680336477933568
2047    886736880519319552
2048    886983233522544640
2049    887101392804085760
2050    887343217045368832
2

In [8]:
# Checking error_list
error_list

[802247111496568832,
 837012587749474308,
 842892208864923648,
 861769973181624320,
 873697596434513921,
 888202515573088257]

In [11]:
# lengh of df_list
print("The lengh of the result", len(df_list))

The lengh of the result 2069


In [12]:
# We repeat the same operation for the tweet_ids that we coudln't get and append the result to df_list
ee_list = []
for e in error_list:
    try:
        favorite_count = page['favorite_count']
        retweet_count = page['retweet_count']
        date_time = page['created_at']     
        df_list.append({'tweet_id': int(tweet_id),
                        'favorite_count': int(favorite_count),
                        'retweet_count': int(retweet_count),
                        'date_time': pd.to_datetime(date_time)})
        
    except Exception:
        print(str(tweet_id)+ " _ " + str(e))
        ee_list.append(e)

0       666020888022790149
1       666029285002620928
2       666033412701032449
3       666044226329800704
4       666049248165822465
5       666050758794694657
6       666051853826850816
7       666055525042405380
8       666057090499244032
9       666058600524156928
10      666063827256086533
11      666071193221509120
12      666073100786774016
13      666082916733198337
14      666094000022159362
15      666099513787052032
16      666102155909144576
17      666104133288665088
18      666268910803644416
19      666273097616637952
20      666287406224695296
21      666293911632134144
22      666337882303524864
23      666345417576210432
24      666353288456101888
25      666362758909284353
26      666373753744588802
27      666396247373291520
28      666407126856765440
29      666411507551481857
               ...        
2045    886366144734445568
2046    886680336477933568
2047    886736880519319552
2048    886983233522544640
2049    887101392804085760
2050    887343217045368832
2

In [13]:
# New lengh of df_list
print("The lengh of the result after querying the initial errors", len(df_list))

The lengh of the result after querying the initial errors 2069


In [14]:
# Create dataframe from the available list of dicitionaries
json_tweets = pd.DataFrame(df_list, columns = ['tweet_id', 'favorite_count', 'retweet_count', 'date_time'])

# Save the dataFrame in file
json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [15]:
# Read the saved tweet_json.txt file into a dataframe
json_df = pd.read_csv('tweet_json.txt', encoding = 'utf-8')
json_df

tweet_id  favorite_count  retweet_count            date_time
0     666020888022790149            2558            521  2015-11-15 22:32:08
1     666029285002620928             131             47  2015-11-15 23:05:30
2     666033412701032449             125             45  2015-11-15 23:21:54
3     666044226329800704             302            141  2015-11-16 00:04:52
4     666049248165822465             109             40  2015-11-16 00:24:50
5     666050758794694657             133             58  2015-11-16 00:30:50
6     666051853826850816            1235            860  2015-11-16 00:35:11
7     666055525042405380             441            257  2015-11-16 00:49:46
8     666057090499244032             300            142  2015-11-16 00:55:59
9     666058600524156928             114             58  2015-11-16 01:01:59
10    666063827256086533             480            223  2015-11-16 01:22:45
11    666071193221509120             148             62  2015-11-16 01:52:02
12    666073100786774016             326            169  2015-11-16 01:59:36
13    666082916733198337             120             45  2015-11-16 02:38:37
14    666094000022159362             166             76  2015-11-16 03:22:39
15    666099513787052032             158             71  2015-11-16 03:44:34
16    666102155909144576              80             13  2015-11-16 03:55:04
17    666104133288665088           14465           6699  2015-11-16 04:02:55
18    666268910803644416             105             36  2015-11-16 14:57:41
19    666273097616637952             178             77  2015-11-16 15:14:19
20    666287406224695296             151             68  2015-11-16 16:11:11
21    666293911632134144             513            359  2015-11-16 16:37:02
22    666337882303524864             200             94  2015-11-16 19:31:45
23    666345417576210432             301            142  2015-11-16 20:01:42
24    666353288456101888             225             74  2015-11-16 20:32:58
25    666362758909284353             786            579  2015-11-16 21:10:36
26    666373753744588802             191             95  2015-11-16 21:54:18
27    666396247373291520             169             87  2015-11-16 23:23:41
28    666407126856765440             111             41  2015-11-17 00:06:54
29    666411507551481857             451            333  2015-11-17 00:24:19
...                  ...             ...            ...                  ...
2039  886258384151887873           28091           6385  2017-07-15 16:17:19
2040  886366144734445568           21213           3229  2017-07-15 23:25:31
2041  886680336477933568           22531           4532  2017-07-16 20:14:00
2042  886736880519319552           12130           3341  2017-07-16 23:58:41
2043  886983233522544640           35327           7871  2017-07-17 16:17:36
2044  887101392804085760           30687           6041  2017-07-18 00:07:08
2045  887343217045368832           33832          10543  2017-07-18 16:08:03
2046  887473957103951883           69440          18475  2017-07-19 00:47:34
2047  887517139158093824           46430          11818  2017-07-19 03:39:09
2048  887705289381826560           30316           5465  2017-07-19 16:06:48
2049  888078434458587136           21859           3551  2017-07-20 16:49:33
2050  888554962724278272           20015           3644  2017-07-22 00:23:06
2051  888804989199671297           25706           4416  2017-07-22 16:56:37
2052  888917238123831296           29214           4565  2017-07-23 00:22:39
2053  889278841981685760           25403           5519  2017-07-24 00:19:32
2054  889531135344209921           15154           2262  2017-07-24 17:02:04
2055  889638837579907072           27299           4604  2017-07-25 00:10:02
2056  889665388333682689           48283          10196  2017-07-25 01:55:32
2057  889880896479866881           27913           5033  2017-07-25 16:11:53
2058  890006608113172480           30784           7439  2017-07-26 00:31:25
2059  890240

# Assess
After gathering data, our next step is to clean this data. We will be asessing the data both visually and programmatically

In [16]:
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [17]:
archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [18]:
archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [19]:
archive['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [20]:
archive['rating_denominator'].value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [21]:
archive['name'].value_counts()

None         745
a             55
Charlie       12
Cooper        11
Oliver        11
Lucy          11
Penny         10
Lola          10
Tucker        10
Bo             9
Winston        9
the            8
Sadie          8
Buddy          7
Toby           7
Daisy          7
Bailey         7
an             7
Stanley        6
Oscar          6
Bella          6
Milo           6
Jax            6
Rusty          6
Koda           6
Dave           6
Leo            6
Jack           6
Scout          6
very           5
            ... 
Kramer         1
Kaia           1
Maxwell        1
Perry          1
Harnold        1
Banditt        1
Pilot          1
Barney         1
Skittles       1
Pip            1
Alexander      1
Aja            1
Rilo           1
Tripp          1
Lilah          1
Kulet          1
Strudel        1
Mutt           1
Butters        1
Milky          1
Halo           1
Blipson        1
Kathmandu      1
Bayley         1
Pippin         1
Gunner         1
Juckson        1
Obie          

In [22]:
archive['doggo'].value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [23]:
archive['floofer'].value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [24]:
archive['pupper'].value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [25]:
archive['puppo'].value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [26]:
images[images.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [27]:
images[images.jpg_url.duplicated()]

tweet_id                                            jpg_url  \
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
1364  761371037149827077  https://pbs.twimg.com/tweet_video_thumb/CeBym7...   
1368  761750502866649088    https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
1387  766078092750233600    https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg   
1407  770093767776997377    https://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg   
1417  771171053431250945    https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
1427  772615324260794368    https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg   
1446  775898661951791106    https://pbs.twimg.com/media/CiyHLocU4AI2pJu.jpg   
1453  776819012571455488    https://pbs.twimg.com/media/CW88XN4WsAAlo8r.jpg   
1456  777641927919427584    https://pbs.twimg.com/media/CmoPdmHW8AAi8BI.jpg   
1463  778396591732486144    https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg   
1476  780496263422808064    https://pbs.twimg.com/media/Ck2d7tJWUAEPTL3.jpg   
1487  782021823840026624    https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1495  783347506784731136    https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
1510  786036967502913536    https://pbs.twimg.com/media/CtKHLuCWYAA2TTs.jpg   
1522  788070120937619456    https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg   
1538  790723298204217344    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1541  791026214425268224    https://pbs.twimg.com/media/CpmyNumW8AAAJGj.jpg   
1564  793614319594401792    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1569  794355576146903043    https://pbs.twimg.com/media/CvJCabcWgAIoUxW.jpg   
1571  794983741416415232    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1579  796177847564038144    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1588  798340744599797760    https://pbs.twimg.com/media/CrXhIqBW8AA6Bse.jpg   
1589  798628517273620480    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
1590  798644042770751489    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
1591  798665375516884993    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
...                  ...                                                ...   
1619  802624713319034886    https://pbs.twimg.com/media/CsrjryzWgAAZY00.jpg   
1624  803692223237865472    https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   
1627  804413760345620481    https://pbs.twimg.com/media/CuRDF-XWcAIZSer.jpg   
1634  805958939288408065    https://pbs.twimg.com/media/CtzKC7zXEAALfSo.jpg   
1636  806242860592926720    https://pbs.twimg.com/media/Ct72q9jWcAAhlnw.jpg   
1640  807059379405148160    https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg   
1645  808134635716833280    https://pbs.twimg.com/media/Cx5R8wPVEAALa9r.jpg   
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1683  813944609378369540    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1693  816014286006976512    https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1712  818588835076603904    https://pbs.twimg.com/media/Crwxb5yWgAAX5P_.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1727  820446719150292993    https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
1736  821813639212650496    https://pbs.twimg.com/media/CtVAvX-WIAAcGTf.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEA

In [28]:
images['jpg_url'].value_counts()

https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg                                            2
https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg                                            2
https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg                                            2
https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg                                            2
https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg                                            2
https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg                                            2
https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg                                            2
https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg                                            2
https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg                                            2
https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg                                            2
https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg                       

In [29]:
json_df['tweet_id'].value_counts()

685532292383666176    1
826598365270007810    1
680494726643068929    1
692158366030913536    1
714606013974974464    1
715696743237730304    1
776477788987613185    1
772114945936949249    1
699775878809702401    1
780858289093574656    1
700462010979500032    1
732726085725589504    1
738883359779196928    1
798644042770751489    1
743510151680958465    1
833722901757046785    1
668620235289837568    1
842765311967449089    1
685315239903100929    1
673686845050527744    1
680473011644985345    1
666051853826850816    1
675853064436391936    1
693231807727280129    1
705475953783398401    1
829449946868879360    1
759923798737051648    1
667160273090932737    1
680934982542561280    1
743545585370791937    1
                     ..
881536004380872706    1
825026590719483904    1
834209720923721728    1
775733305207554048    1
669564461267722241    1
879492040517615616    1
720775346191278080    1
666362758909284353    1
750506206503038976    1
693155686491000832    1
7936017773084631

In [30]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2069 entries, 0 to 2068
Data columns (total 4 columns):
tweet_id          2069 non-null int64
favorite_count    2069 non-null int64
retweet_count     2069 non-null int64
date_time         2069 non-null object
dtypes: int64(3), object(1)
memory usage: 80.8+ KB


In [31]:
images.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 168.2+ KB


In [32]:
archive['retweeted_status_timestamp']

0                             NaN
1                             NaN
2                             NaN
3                             NaN
4                             NaN
5                             NaN
6                             NaN
7                             NaN
8                             NaN
9                             NaN
10                            NaN
11                            NaN
12                            NaN
13                            NaN
14                            NaN
15                            NaN
16                            NaN
17                            NaN
18                            NaN
19      2017-07-19 00:47:34 +0000
20                            NaN
21                            NaN
22                            NaN
23                            NaN
24                            NaN
25                            NaN
26                            NaN
27                            NaN
28                            NaN
29            

# Quality 
## Accuracy, Validity, Completeness, Consistency

### Archive Dataset
- retweeted_status_timestamp, timestamp are object (string) instead of datetime
- In several columns null objects are non-null (None to NaN)
- Many invalid values in name columns like a, None, an
- Some columns types are float instaed of string (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id and tweet_id); String is desirable because we don't want any operations on them

### Image Dataset
- Missing values from images dataset (2068 rows instead of 2356)
- Many tweet_ids have same jpg_url
- tweet_id is an integer instead of string

### JSON Dataset
- date_time is an object(string) instead of datetime

# Tidiness
## Structural issues
- dog_type can be made into one column
- two separate dataframes can be created with tweet_id as primary key; one which contains details of tweet and the other which contains details of dogs
- We dont need to show different predictions and confidence intervals. All the AI based infoarmation can be condense based on true values
- All three datasets are about the same tweets. Therefore, it is logival to creat one master data based on tweet id af all the datasets

# Clean
We will now define the cleaning process and actually do the cleaning based on the insights gathered from our assessment step

In [33]:
# Create copies of all the dataframes as backup
archive_clean = archive.copy()
images_clean = images.copy()
json_df_clean = json_df.copy()

## Definition
- use pd.todatetime to convert retweeted_status_timestamp and timestamp to datetime

In [34]:
archive_clean['retweeted_status_timestamp'] = pd.to_datetime(archive_clean['retweeted_status_timestamp'])
archive_clean['timestamp'] = pd.to_datetime(archive_clean['timestamp'])

In [35]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null datetime64[ns]
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime

## Definition
- Use numpy to convert None to nan

In [37]:
# Cleaning rows and columsn which we willl not require
# Some NaN values were not being read as NaN so I converted it using numpy
for c in archive_clean.select_dtypes(include=["object"]).columns:
    archive_clean[c] = archive_clean[c].replace('None', np.nan)

archive_clean.info()
# archive_clean.replace('None', np.nan, inplace=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null datetime64[ns]
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          1611 non-null object
doggo                         97 non-null object
floofer                       10 non-null object
pupper                        257 non-null object
puppo                         30 non-null object
dtypes: datetime64[ns](

In [38]:
archive_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [39]:
# Delete the retweets
archive_clean = archive_clean[pd.isnull(archive_clean.retweeted_status_id)]
archive_clean.info()
# Delete tweets with no pictures
# archive_clean = archive_clean.dropna(subset = ['jpg_url'])

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null datetime64[ns]
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null datetime64[ns]
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          1495 non-null object
doggo                         87 non-null object
floofer                       10 non-null object
pupper                        234 non-null object
puppo                         25 non-null object
dtypes: datetime64[ns](2), fl

In [40]:
# Select the columns to melt and to remain
columns_to_melt = ['doggo', 'floofer', 'pupper', 'puppo']
columns_to_keep = [x for x in archive_clean.columns.tolist() if x not in columns_to_melt]

# Mlet the the columns into values
archive_clean = pd.melt(archive_clean, id_vars = columns_to_keep, value_vars = columns_to_melt, 
                         var_name = 'stages', value_name = 'dog_stage')

# Delete column 'stages'
archive_clean = archive_clean.drop('stages', 1)

# Filter for unique values then remove duplicate values based on 'dog_stage' values

# This part for test *
print(archive_clean.dog_stage.value_counts())

archive_clean = archive_clean.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

# This part for test
# print(archive_clean.dog_stage.value_counts())
# print(len(archive_clean))
archive_clean.dog_stage.value_counts()

pupper     234
doggo       87
puppo       25
floofer     10
Name: dog_stage, dtype: int64


Series([], Name: dog_stage, dtype: int64)

In [41]:
# Loop on all the texts and check if the comment has one of the above conditions
# and append the result in a list
dog_names = []

for text in archive_clean['text']:
    # Start with 'This is ' and the fisrt letter of the name is uppercase
    if text.startswith('This is ') and re.match(r'[A-Z].*', text.split()[2]):
        dog_names.append(text.split()[2].strip(',').strip('.'))
    # Start with 'Meet ' and the fisrt letter of the name is uppercase
    elif text.startswith('Meet ') and re.match(r'[A-Z].*', text.split()[1]):
        dog_names.append(text.split()[1].strip(',').strip('.'))
    # Start with 'Say hello to ' and the fisrt letter of the name is uppercase
    elif text.startswith('Say hello to ') and re.match(r'[A-Z].*', text.split()[3]):
        dog_names.append(text.split()[3].strip(',').strip('.'))
    # Start with 'Here we have ' and the fisrt letter of the name is uppercase
    elif text.startswith('Here we have ') and re.match(r'[A-Z].*', text.split()[3]):
        dog_names.append(text.split()[3].strip(',').strip('.'))
    # Contain 'named' and the fisrt letter of the name is uppercase
    elif 'named' in text and re.match(r'[A-Z].*', text.split()[text.split().index('named') + 1]):
        dog_names.append(text.split()[text.split().index('named') + 1].strip(',').strip('.'))
    # No name specified or other style
    else:
        dog_names.append('NaN')

# Test
len(dog_names)

# Save the result in a new column 'dog_name'
archive_clean['dog_name'] = dog_names

# Test
print("New column dog_name count \n", archive_clean.dog_name.value_counts())
print("Old column name count \n", archive_clean.name.value_counts())

New column dog_name count 
 NaN          765
Charlie       11
Lucy          11
Oliver        10
Cooper        10
Tucker         9
Penny          9
Sadie          8
Lola           8
Winston        8
Daisy          7
Jax            6
Bailey         6
Koda           6
Oscar          6
Toby           6
Stanley        6
Bo             6
Bella          6
Leo            5
Buddy          5
Chester        5
Bentley        5
Louis          5
Milo           5
Rusty          5
Scout          5
Dexter         4
Phil           4
Chip           4
            ... 
Carll          1
Loomis         1
Spark          1
Bayley         1
Halo           1
Pupcasso       1
Pilot          1
Lillie         1
Jackie         1
Mojo           1
BeBe           1
Henry          1
Pumpkin        1
Tycho          1
Alejandro      1
Cedrick        1
Tebow          1
Skittles       1
Milky          1
Tobi           1
Pip            1
Alexander      1
Aja            1
Rilo           1
Lilah          1
Kulet          1
Str

In [42]:
# There are some names in name column which was not deduced by our above code. Therefore, I am chevcking whether dog_name column has NaN values
# If NaN values found then I look in the name column for a value
archive_clean['dog_name'] = np.where(archive_clean.dog_name.isnull(),
                                     archive_clean['name'], 
                                     archive_clean['dog_name'])

In [43]:
# Drop the unwanted columns
archive_clean = archive_clean.drop(['name'], 1)
archive_clean = archive_clean.drop(['retweeted_status_id'], 1)
archive_clean = archive_clean.drop(['retweeted_status_user_id'], 1)
archive_clean = archive_clean.drop(['retweeted_status_timestamp'], 1)
archive_clean = archive_clean.drop(['dog_stage'], 1)
archive_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
4362  889665388333682689                    NaN                  NaN   
4364  889531135344209921                    NaN                  NaN   
4417  878776093423087618                    NaN                  NaN   
4436  874012996292530176                    NaN                  NaN   
4465  867421006826221569                    NaN                  NaN   
4496  859607811541651456                    NaN                  NaN   
4515  855851453814013952                    NaN                  NaN   
4684  825535076884762624                    NaN                  NaN   
4686  825026590719483904                    NaN                  NaN   
4698  822872901745569793                    NaN                  NaN   
4716  819952236453363712                    NaN                  NaN   
4809  803773340896923648                    NaN                  NaN   
4818  802239329049477120                    NaN                  NaN   
4868  793195938047070209                    NaN                  NaN   
4884  790946055508652032                    NaN                  NaN   
4904  787717603741622272                    NaN                  NaN   
4914  786051337297522688           7.727430e+17         7.305050e+17   
4925  784183165795655680                    NaN                  NaN   
4946  780931614150983680                    NaN                  NaN   
5101  756275833623502848                    NaN                  NaN   
5123  752519690950500352                    NaN                  NaN   
5136  751132876104687617                    NaN                  NaN   
5208  744995568523612160                    NaN                  NaN   
5220  743253157753532416                    NaN                  NaN   
5255  738537504001953792                    NaN                  NaN   
6525  892420643555336193                    NaN                  NaN   
6526  892177421306343426                    NaN                  NaN   
6527  891815181378084864                    NaN                  NaN   
6528  891689557279858688                    NaN                  NaN   
6529  891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
8670  666411507551481857                    NaN                  NaN   
8671  666407126856765440                    NaN                  NaN   
8672  666396247373291520                    NaN                  NaN   
8673  666373753744588802                    NaN                  NaN   
8674  666362758909284353                    NaN                  NaN   
8675  666353288456101888                    NaN                  NaN   
8676  666345417576210432                    NaN                  NaN   
8677  666337882303524864                    NaN                  NaN   
8678  666293911632134144                    NaN                  NaN   
8679  666287406224695296                    NaN                  NaN   
8680  666273097616637952                    NaN                  NaN   
8681  666268910803644416                    NaN                  NaN   
8682  666104133288665088                    NaN                  NaN   
8683  666102155909144576                    NaN                  NaN   
8684  666099513787052032                    NaN                  NaN   
8685  666094000022159362                    NaN                  NaN   
8686  666082916733198337                    NaN                  NaN   
8687  666073100786774016                    NaN                  NaN   
8688  666071193221509120                    NaN                  NaN   
8689  666063827256086533                    NaN                  NaN   
8690  666058600524156928                    NaN                  NaN   
8691  666057090499244032                    NaN                  NaN   
8692  666055525042405380                    NaN                  NaN   
8693  666051853826850816                    NaN         

## Definition
### Problem
- Some columns types are float instaed of string (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id and tweet_id); String is desirable because we don't want any operations on them

### Solution
- change datatype to string

In [44]:
archive_clean.in_reply_to_status_id = archive_clean.in_reply_to_status_id.apply(str)
archive_clean.in_reply_to_user_id = archive_clean.in_reply_to_user_id.apply(str)
archive_clean.retweeted_status_id = archive_clean.retweeted_status_id.apply(str)
archive_clean.retweeted_status_user_id = archive_clean.retweeted_status_user_id.apply(str)
archive_clean.tweet_id = archive_clean.tweet_id.apply(str)

AttributeError: 'DataFrame' object has no attribute 'retweeted_status_id'

In [ ]:
archive_clean.info()

## Definition
### Problem
- Missing values from images dataset (2075 rows instead of 2175)

### Solution
- 

In [45]:
# next let's condense the dog breed analysis
breed = []
confidence = []

def breed_confidence(row):
    if row['p1_dog'] == True:
        breed.append(row['p1'])
        confidence.append(row['p1_conf'])
    elif row['p2_dog'] == True:
        breed.append(row['p2'])
        confidence.append(row['p2_conf'])
    elif row['p3_dog'] == True:
        breed.append(row['p3'])
        confidence.append(row['p3_conf'])
    else:
        breed.append('Unidentifiable')
        confidence.append(0)
      
 # this isn't about the parsing AI so the rest of this data is deletable

images_clean.apply(breed_confidence, axis=1)
images_clean['breed'] = breed
images_clean['confidence'] = confidence
images_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074   True              collie   
1        1                 redbone  0.506826   True  miniature_pinscher   
2        1         German_shepherd  0.596461   True            malinois   
3        1     Rhodesian_ridgeback  0.408143   True             redbone   
4        1      miniature_pinscher  0.560311   True          Rottweiler   

    p2_conf p2_dog                   p3   p3_conf p3_dog  \
0  0.156665   True    Shetland_sheepdog  0.061428   True   
1  0.074192   True  Rhodesian_ridgeback  0.072010   True   
2  0.138584   True           bloodhound  0.116197   True   
3  0.360687   True   miniature_pinscher  0.222752   True   
4  0.243682   True             Doberman  0.154629   True   

                    breed  confidence  
0  Welsh_springer_spaniel    0.465074  
1                 redbone    0.506826  
2         German_shepherd    0.596461  
3     Rhodesian_ridgeback    0.408143  
4      miniature_pinscher    0.560311

In [46]:
images_clean.drop(['p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog',], axis=1, inplace=True)
images_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                   breed  confidence  
0        1  Welsh_springer_spaniel    0.465074  
1        1                 redbone    0.506826  
2        1         German_shepherd    0.596461  
3        1     Rhodesian_ridgeback    0.408143  
4        1      miniature_pinscher    0.560311

In [47]:
rates = []

extract_rates = lambda x: rates.append(re.findall(r'(\d+(\.\d+)|(\d+))\/(\d+0)', x, flags=0))

archive_clean['text'].apply(extract_rates)

numerator = []
dog_count = []

for item in rates:
    
    # for tweets with no rating, but a picture, so a dog
    if len(item) == 0:
        numerator.append('NaN')
        dog_count.append(1)
        
    # for tweetss with one rating and one dog
    elif len(item) == 1 and item[0][-1] == '10':
        numerator.append(float(item[0][0]))
        dog_count.append(1)
   
    # for group ratings
    elif len(item) == 1: 
        avg = float(item[0][0]) / (float(item[0][-1]) / 10) 
        numerator.append(avg)
        dog_count.append(float(item[0][-1]) / 10)
   
    # for tweets with more than one rating
    elif len(item) > 1: 
        total = 0
        list = []
        for i in range(len(item)):
            if item[i][-1] == '10': #one tweet has the phrase '50/50' so I'm coding to exclude it
                list.append(item[i])
        for rate in list:
            total = total + float(rate[0])
        avg = total / len(item)
        numerator.append(avg)
        dog_count.append(len(item))
   
    # in order to catch bugs
    else:
        numerator.append('Not parsed')
        dog_count.append('Not parsed') 
        
archive_clean['rating'] = numerator # not need to also add denominator since they are all 10!
archive_clean['dog_count'] = dog_count
archive_clean['rating'].value_counts()

12.0                 503
10.0                 434
11.0                 428
13.0                 307
9.0                  157
8.0                   97
7.0                   54
14.0                  44
5.0                   32
6.0                   32
3.0                   19
4.0                   14
2.0                   10
1.0                    7
8.5                    4
9.5                    4
6.5                    3
7.5                    3
NaN                    3
5.5                    2
15.0                   2
10.5                   2
420.0                  2
0.0                    2
1776.0                 1
9.666666666666666      1
4.5                    1
9.75                   1
11.27                  1
13.5                   1
666.0                  1
182.0                  1
11.5                   1
11.26                  1
Name: rating, dtype: int64

In [48]:
# We can see that all the rating sare below 14 except the joke ratings of 420 and 1776. So we can drop the rating_numerator and rating_denominator columns
archive_clean.drop([ 'rating_numerator', 'rating_denominator'], axis=1, inplace=True)

In [49]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 4362 to 8699
Data columns (total 10 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    2175 non-null object
in_reply_to_user_id      2175 non-null object
timestamp                2175 non-null datetime64[ns]
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
dog_name                 2175 non-null object
rating                   2175 non-null object
dog_count                2175 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1), object(7)
memory usage: 186.9+ KB


In [50]:
archive_clean['dog_count'].value_counts()

1     2135
2       27
8        2
5        2
17       1
15       1
13       1
12       1
11       1
9        1
7        1
4        1
3        1
Name: dog_count, dtype: int64

In [51]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 4362 to 8699
Data columns (total 10 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    2175 non-null object
in_reply_to_user_id      2175 non-null object
timestamp                2175 non-null datetime64[ns]
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
dog_name                 2175 non-null object
rating                   2175 non-null object
dog_count                2175 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1), object(7)
memory usage: 186.9+ KB


In [52]:
json_df_clean

tweet_id  favorite_count  retweet_count            date_time
0     666020888022790149            2558            521  2015-11-15 22:32:08
1     666029285002620928             131             47  2015-11-15 23:05:30
2     666033412701032449             125             45  2015-11-15 23:21:54
3     666044226329800704             302            141  2015-11-16 00:04:52
4     666049248165822465             109             40  2015-11-16 00:24:50
5     666050758794694657             133             58  2015-11-16 00:30:50
6     666051853826850816            1235            860  2015-11-16 00:35:11
7     666055525042405380             441            257  2015-11-16 00:49:46
8     666057090499244032             300            142  2015-11-16 00:55:59
9     666058600524156928             114             58  2015-11-16 01:01:59
10    666063827256086533             480            223  2015-11-16 01:22:45
11    666071193221509120             148             62  2015-11-16 01:52:02
12    666073100786774016             326            169  2015-11-16 01:59:36
13    666082916733198337             120             45  2015-11-16 02:38:37
14    666094000022159362             166             76  2015-11-16 03:22:39
15    666099513787052032             158             71  2015-11-16 03:44:34
16    666102155909144576              80             13  2015-11-16 03:55:04
17    666104133288665088           14465           6699  2015-11-16 04:02:55
18    666268910803644416             105             36  2015-11-16 14:57:41
19    666273097616637952             178             77  2015-11-16 15:14:19
20    666287406224695296             151             68  2015-11-16 16:11:11
21    666293911632134144             513            359  2015-11-16 16:37:02
22    666337882303524864             200             94  2015-11-16 19:31:45
23    666345417576210432             301            142  2015-11-16 20:01:42
24    666353288456101888             225             74  2015-11-16 20:32:58
25    666362758909284353             786            579  2015-11-16 21:10:36
26    666373753744588802             191             95  2015-11-16 21:54:18
27    666396247373291520             169             87  2015-11-16 23:23:41
28    666407126856765440             111             41  2015-11-17 00:06:54
29    666411507551481857             451            333  2015-11-17 00:24:19
...                  ...             ...            ...                  ...
2039  886258384151887873           28091           6385  2017-07-15 16:17:19
2040  886366144734445568           21213           3229  2017-07-15 23:25:31
2041  886680336477933568           22531           4532  2017-07-16 20:14:00
2042  886736880519319552           12130           3341  2017-07-16 23:58:41
2043  886983233522544640           35327           7871  2017-07-17 16:17:36
2044  887101392804085760           30687           6041  2017-07-18 00:07:08
2045  887343217045368832           33832          10543  2017-07-18 16:08:03
2046  887473957103951883           69440          18475  2017-07-19 00:47:34
2047  887517139158093824           46430          11818  2017-07-19 03:39:09
2048  887705289381826560           30316           5465  2017-07-19 16:06:48
2049  888078434458587136           21859           3551  2017-07-20 16:49:33
2050  888554962724278272           20015           3644  2017-07-22 00:23:06
2051  888804989199671297           25706           4416  2017-07-22 16:56:37
2052  888917238123831296           29214           4565  2017-07-23 00:22:39
2053  889278841981685760           25403           5519  2017-07-24 00:19:32
2054  889531135344209921           15154           2262  2017-07-24 17:02:04
2055  889638837579907072           27299           4604  2017-07-25 00:10:02
2056  889665388333682689           48283          10196  2017-07-25 01:55:32
2057  889880896479866881           27913           5033  2017-07-25 16:11:53
2058  890006608113172480           30784           7439  2017-07-26 00:31:25
2059  890240

In [53]:
json_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2069 entries, 0 to 2068
Data columns (total 4 columns):
tweet_id          2069 non-null int64
favorite_count    2069 non-null int64
retweet_count     2069 non-null int64
date_time         2069 non-null object
dtypes: int64(3), object(1)
memory usage: 80.8+ KB


In [55]:
df_master = pd.merge(archive_clean, images_clean, how = 'left', on = ['tweet_id'] )
df_master = pd.merge(df_master, json_df_clean, how = 'left', on = ['tweet_id'])
df_master.to_csv('df_master.csv', encoding = 'utf-8')
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2174
Data columns (total 17 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    2175 non-null object
in_reply_to_user_id      2175 non-null object
timestamp                2175 non-null datetime64[ns]
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
dog_name                 2175 non-null object
rating                   2175 non-null object
dog_count                2175 non-null float64
jpg_url                  1994 non-null object
img_num                  1994 non-null float64
breed                    1994 non-null object
confidence               1994 non-null float64
favorite_count           1994 non-null float64
retweet_count            1994 non-null float64
date_time                1994 non-null object
dtypes: datetime64[ns](1), float64(5), int64(1), object(10)
memory usage: 305.9+ KB


In [59]:
df_master = df_master.dropna(subset = ['jpg_url'])
df_master

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0     889665388333682689                   nan                 nan   
1     889531135344209921                   nan                 nan   
2     878776093423087618                   nan                 nan   
3     874012996292530176                   nan                 nan   
4     867421006826221569                   nan                 nan   
5     859607811541651456                   nan                 nan   
6     855851453814013952                   nan                 nan   
7     825535076884762624                   nan                 nan   
8     825026590719483904                   nan                 nan   
9     822872901745569793                   nan                 nan   
10    819952236453363712                   nan                 nan   
11    803773340896923648                   nan                 nan   
12    802239329049477120                   nan                 nan   
13    793195938047070209                   nan                 nan   
14    790946055508652032                   nan                 nan   
15    787717603741622272                   nan                 nan   
18    780931614150983680                   nan                 nan   
19    756275833623502848                   nan                 nan   
20    752519690950500352                   nan                 nan   
21    751132876104687617                   nan                 nan   
22    744995568523612160                   nan                 nan   
23    743253157753532416                   nan                 nan   
24    738537504001953792                   nan                 nan   
25    892420643555336193                   nan                 nan   
26    892177421306343426                   nan                 nan   
27    891815181378084864                   nan                 nan   
28    891689557279858688                   nan                 nan   
29    891327558926688256                   nan                 nan   
30    891087950875897856                   nan                 nan   
31    890971913173991426                   nan                 nan   
...                  ...                   ...                 ...   
2145  666411507551481857                   nan                 nan   
2146  666407126856765440                   nan                 nan   
2147  666396247373291520                   nan                 nan   
2148  666373753744588802                   nan                 nan   
2149  666362758909284353                   nan                 nan   
2150  666353288456101888                   nan                 nan   
2151  666345417576210432                   nan                 nan   
2152  666337882303524864                   nan                 nan   
2153  666293911632134144                   nan                 nan   
2154  666287406224695296                   nan                 nan   
2155  666273097616637952                   nan                 nan   
2156  666268910803644416                   nan                 nan   
2157  666104133288665088                   nan                 nan   
2158  666102155909144576                   nan                 nan   
2159  666099513787052032                   nan                 nan   
2160  666094000022159362                   nan                 nan   
2161  666082916733198337                   nan                 nan   
2162  666073100786774016                   nan                 nan   
2163  666071193221509120                   nan                 nan   
2164  666063827256086533                   nan                 nan   
2165  666058600524156928                   nan                 nan   
2166  666057090499244032                   nan                 nan   
2167  666055525042405380                   nan                 nan   
2168  666051853826850816                   nan                 nan   
2169  666050758794694657                   nan                 nan   
2170  666049248165822465  

In [60]:
# Delete column date_time we imported from the API, it has the same values as timestamp column
df_master = df_master.drop('date_time', 1)
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2174
Data columns (total 16 columns):
tweet_id                 1994 non-null int64
in_reply_to_status_id    1994 non-null object
in_reply_to_user_id      1994 non-null object
timestamp                1994 non-null datetime64[ns]
source                   1994 non-null object
text                     1994 non-null object
expanded_urls            1994 non-null object
dog_name                 1994 non-null object
rating                   1994 non-null object
dog_count                1994 non-null float64
jpg_url                  1994 non-null object
img_num                  1994 non-null float64
breed                    1994 non-null object
confidence               1994 non-null float64
favorite_count           1994 non-null float64
retweet_count            1994 non-null float64
dtypes: datetime64[ns](1), float64(5), int64(1), object(9)
memory usage: 264.8+ KB


In [62]:
for c in df_master.select_dtypes(include=["object"]).columns:
    df_master[c] = df_master[c].replace('None', np.nan)

for c in df_master.select_dtypes(include=["object"]).columns:
    df_master[c] = df_master[c].replace('nan', np.nan)
    
for c in df_master.select_dtypes(include=["object"]).columns:
    df_master[c] = df_master[c].replace('NaN', np.nan)

In [64]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2174
Data columns (total 16 columns):
tweet_id                 1994 non-null int64
in_reply_to_status_id    23 non-null object
in_reply_to_user_id      23 non-null object
timestamp                1994 non-null datetime64[ns]
source                   1994 non-null object
text                     1994 non-null object
expanded_urls            1994 non-null object
dog_name                 1369 non-null object
rating                   1993 non-null float64
dog_count                1994 non-null float64
jpg_url                  1994 non-null object
img_num                  1994 non-null float64
breed                    1994 non-null object
confidence               1994 non-null float64
favorite_count           1994 non-null float64
retweet_count            1994 non-null float64
dtypes: datetime64[ns](1), float64(6), int64(1), object(8)
memory usage: 264.8+ KB


In [65]:
df_master.to_csv('twitter_archive_master.csv', index=False, encoding = 'utf-8')